feat(ci): Add multi-arch Docker image build support for clp-package (amd64 + arm64).

junhaoliao opened the pull request:
<!-- markdownlint-disable MD012 -->

# Description

Add multi-architecture (amd64 and arm64) support for the `clp-package` Docker image build and push workflow. Arm64 builds are restricted to push-to-main only to save CI resources on PRs.

**Changes:**

1. **`.github/actions/clp-build-runtime-image/action.yaml`:**
   - Add `arch` input parameter to specify target architecture (default: `amd64`)
   - Add `map-arch` step that maps `amd64`/`arm64` to Docker platform `linux/amd64`/`linux/arm64`, with validation for unsupported values
   - Update `docker/build-push-action` to use `platforms` from the mapping step
   - Add `provenance: false` to create simple images instead of manifest lists, allowing `docker manifest create` to combine the per-arch images later
   - Use `docker/metadata-action`'s `flavor.suffix` to append architecture suffix to all image tags (e.g., `clp-package:main-amd64`), fixing a bug where manual concatenation only suffixed the last tag

2. **`.github/workflows/clp-artifact-build.yaml`:**
   - Add `ubuntu-jammy-aarch64-deps-image` job that builds deps on arm64 runners (only on push to main)
   - Rename `ubuntu-jammy-deps-image` to `ubuntu-jammy-x86_64-deps-image` for consistency with manylinux/musllinux arch naming
   - Add `package-image` matrix strategy for `["amd64", "arm64"]` on push to main, `["amd64"]` on PRs
   - Add `package-image-multiarch-manifest` job that creates and pushes a combined multi-arch manifest for the branch tag (`main`) on every push, and additionally for the `nightly` tag on scheduled runs
   - Use Docker convention (`amd64`/`arm64`) for published image tags, while keeping `x86`/`aarch64` for internal deps image names

# Checklist

* [x] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [x] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [x] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

- Workflow ran successfully: https://github.com/junhaoliao/clp/actions/runs/20521459673
  - Verified `provenance: false` resolves the "is a manifest list" error when creating multi-arch manifests
- Tested on a Dell XPS 9345 (Snapdragon X Elite X1E-80-100, ARMv8 compatible):
  ```bash
  > uname -a
  Linux yscope-junhao 6.6.87.2-microsoft-standard-WSL2 #1 SMP PREEMPT_DYNAMIC Thu Jun  5 18:31:42 UTC 2025 aarch64 aarch64 aarch64 GNU/Linux
  > docker run --rm ghcr.io/junhaoliao/clp/clp-package:main clp-s --help
  Usage: clp-s [OPTIONS] COMMAND [COMMAND ARGUMENTS]
  COMMAND is one of:
    c - compress
    x - decompress
    s - search

  Try  c --help OR x --help OR s --help for command-specific details.

  General options:
    -h [ --help ]          Print help

  > docker run --rm ghcr.io/junhaoliao/clp/clp-package:main uname -p -i
  aarch64 aarch64
  ```
- Tested on an AMD Ryzen 9 9950X3D (x86_64):
  ```bash
  > docker run --rm ghcr.io/junhaoliao/clp/clp-package:main clp-s --help
  Usage: clp-s [OPTIONS] COMMAND [COMMAND ARGUMENTS]
  COMMAND is one of:
    c - compress
    x - decompress
    s - search

  Try  c --help OR x --help OR s --help for command-specific details.

  General options:
    -h [ --help ]          Print help

  > docker run --rm ghcr.io/junhaoliao/clp/clp-package:main uname -p -i
  x86_64 x86_64
  ```
- Verified multi-arch manifest structure using `docker buildx imagetools inspect`:
  ```bash
  > docker buildx imagetools inspect ghcr.io/junhaoliao/clp/clp-package:main
  Name:      ghcr.io/junhaoliao/clp/clp-package:main
  MediaType: application/vnd.docker.distribution.manifest.list.v2+json
  Digest:    sha256:bd374f5624b13cadb4d243563391ff6220c09a72c784de73fb9e932a5ddf8397
             
  Manifests: 
    Name:      ghcr.io/junhaoliao/clp/clp-package:main@sha256:2a1e392f0a04f7ffe2e0185e8fbb07e80a4ddad6fa27befe2d3b139c7c2cdb82
    MediaType: application/vnd.docker.distribution.manifest.v2+json
    Platform:  linux/amd64
               
    Name:      ghcr.io/junhaoliao/clp/clp-package:main@sha256:71d0b9f9c38dbee01b0f7e5ec6948cfe369e442c25ae901199cf3b66fb5c3b0c
    MediaType: application/vnd.docker.distribution.manifest.v2+json
    Platform:  linux/arm64

  > docker buildx imagetools inspect ghcr.io/junhaoliao/clp/clp-package:nightly
  Name:      ghcr.io/junhaoliao/clp/clp-package:nightly
  MediaType: application/vnd.docker.distribution.manifest.list.v2+json
  Digest:    sha256:bd374f5624b13cadb4d243563391ff6220c09a72c784de73fb9e932a5ddf8397
             
  Manifests: 
    Name:      ghcr.io/junhaoliao/clp/clp-package:nightly@sha256:2a1e392f0a04f7ffe2e0185e8fbb07e80a4ddad6fa27befe2d3b139c7c2cdb82
    MediaType: application/vnd.docker.distribution.manifest.v2+json
    Platform:  linux/amd64
               
    Name:      ghcr.io/junhaoliao/clp/clp-package:nightly@sha256:71d0b9f9c38dbee01b0f7e5ec6948cfe369e442c25ae901199cf3b66fb5c3b0c
    MediaType: application/vnd.docker.distribution.manifest.v2+json
    Platform:  linux/arm64

  > docker buildx imagetools inspect ghcr.io/junhaoliao/clp/clp-package:main-amd64
  Name:      ghcr.io/junhaoliao/clp/clp-package:main-amd64
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Digest:    sha256:2a1e392f0a04f7ffe2e0185e8fbb07e80a4ddad6fa27befe2d3b139c7c2cdb82

  > docker buildx imagetools inspect ghcr.io/junhaoliao/clp/clp-package:main-arm64
  Name:      ghcr.io/junhaoliao/clp/clp-package:main-arm64
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Digest:    sha256:71d0b9f9c38dbee01b0f7e5ec6948cfe369e442c25ae901199cf3b66fb5c3b0c

  > docker buildx imagetools inspect ghcr.io/junhaoliao/clp/clp-package:nightly-amd64
  Name:      ghcr.io/junhaoliao/clp/clp-package:nightly-amd64
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Digest:    sha256:2a1e392f0a04f7ffe2e0185e8fbb07e80a4ddad6fa27befe2d3b139c7c2cdb82

  > docker buildx imagetools inspect ghcr.io/junhaoliao/clp/clp-package:nightly-arm64
  Name:      ghcr.io/junhaoliao/clp/clp-package:nightly-arm64
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Digest:    sha256:71d0b9f9c38dbee01b0f7e5ec6948cfe369e442c25ae901199cf3b66fb5c3b0c
  ```
  Both `:main` and `:nightly` are manifest lists pointing to the same per-arch images. `:main-amd64`/`:nightly-amd64` and `:main-arm64`/`:nightly-arm64` are single-arch images whose digests match the entries in their respective manifest lists.
- Pushed to fork's main branch and the full workflow succeeded (all 22 jobs green, including `package-image-amd64`, `package-image-arm64`, and `package-image-multiarch-manifest`):
  https://github.com/junhaoliao/clp/actions/runs/25506334108
- On the PR-triggered workflow run, both `ubuntu-jammy-aarch64-deps-image` and
  `package-image-multiarch-manifest` are correctly skipped (as expected, since arm64
  builds only run on push to main):
  https://github.com/y-scope/clp/actions/runs/25351763296?pr=1830

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html

junhaoliao committed:
feat(ci): Add support for multi-architecture builds and package image manifest.

junhaoliao committed:
fix(ci): Simplify runner configuration and matrix strategy for ARM64 build.

junhaoliao committed:
fix(ci): Disable provenance for runtime image builds to support multi-arch manifest creation.

junhaoliao committed:
Merge branch 'main' into multi-arch-docker

junhaoliao committed:
Merge branch 'main' into multi-arch-docker

junhaoliao committed:
Merge branch 'main' into multi-arch-docker

jackluo923 commented on .github/workflows/clp-artifact-build.yaml:
```diff
@@ -562,3 +611,36 @@ jobs:
+
+  package-image-multiarch-manifest:
+    name: "package-image-multiarch-manifest"
+    if: "github.event_name != 'pull_request' && github.ref == 'refs/heads/main'"
```
I think we should also check for package-image should be successful
&& needs.package-image.result == 'success'

junhaoliao committed:
fix(ci): Add dependency check for multi-arch manifest packaging step.

junhaoliao committed:
Merge branch 'main' into multi-arch-docker

junhaoliao committed:
Merge branch 'main' into multi-arch-docker

junhaoliao committed:
Merge branch 'main' into multi-arch-docker

Resolve merge conflicts and align naming conventions with origin/main:
- Use DEPS_IMAGE_NAME_PREFIX_X86/AARCH64 (matching main) instead of AMD64/ARM64
- Use &runner anchor (matching main) instead of &runner-amd64
- Add ubuntu-jammy-aarch64-deps-image job (only runs on push to main)
- Restrict arm64/aarch64 package-image builds to push to main only
- Update arch values from amd64/arm64 to x86_64/aarch64 for consistency

junhaoliao commented on .github/actions/clp-build-runtime-image/action.yaml:
```diff
@@ -21,6 +21,10 @@ inputs:
       (e.g. jammy, focal, etc.)"
     required: false
+  arch:
+    description: "Target architecture (amd64 or arm64)"
```
amd64 and arm64 are more commonly used docker image arch names. i propose that for internal toolchain / dependency related images, we can keep the x86_64 / aarch64 naming, but for the user facing images, we should use arch name `amd64` / `arm64`.

jackluo923 commented:
The `docker/metadata-action` step in `.github/actions/clp-build-runtime-image/action.yaml` doesn't specify custom `tags`, so it uses its defaults, which include `type=schedule`. This produces a `nightly` tag on every daily schedule run (which this workflow already has via `cron: "15 0 * * *"`).

Before this PR, that was harmless — `nightly` was just a redundant single-arch tag. But now, both the amd64 and arm64 matrix jobs push `clp-package:nightly` without an arch suffix (the suffix only lands on the last tag in the newline-delimited output), so the second one to finish overwrites the first. The `nightly` tag becomes a single-arch image determined by a race condition.

I'd suggest fixing this at the root by providing explicit tag rules that omit the `type=schedule` rule, since nightly images don't appear to be an intentional feature:

```yaml
- name: "Extract GitHub Metadata"
  id: "extract-gh-meta"
  uses: "docker/metadata-action@902fa8ec7d6ecbf8d84d538b9b233a880e428804"
  with:
    images: "${{inputs.image_registry}}/${{steps.sanitization.outputs.REPOSITORY}}\
      /${{steps.compute-meta.outputs.IMAGE_NAME}}"
    tags: |
      type=ref,event=branch
      type=ref,event=pr
      type=ref,event=tag
```

This produces exactly one tag per event type (e.g., `main` on push, `pr-N` on PRs), so the existing `-${{inputs.arch}}` suffix works correctly without changes, and no unwanted `nightly` images get published. If nightly images are wanted in the future, the `flavor` input with `suffix=-${{inputs.arch}},onlatest=true` should be used instead.

jackluo923 reviewed (changes requested):
Check the top level comment I added.

junhaoliao committed:
fix(ci): Use flavor.suffix to append arch to all Docker image tags

The previous approach of concatenating `-${{inputs.arch}}` to the
metadata action's output only suffixed the last tag in the
newline-delimited list. Tags like `main` and `nightly` were left
unsuffixed, causing both matrix jobs to push to the same tag (race
condition). Use `flavor.suffix` instead, which correctly appends the
arch suffix to every tag.

junhaoliao commented:
> The `docker/metadata-action` step in `.github/actions/clp-build-runtime-image/action.yaml` doesn't specify custom `tags`, so it uses its defaults, which include `type=schedule`. This produces a `nightly` tag on every daily schedule run (which this workflow already has via `cron: "15 0 * * *"`).
> 
> Before this PR, that was harmless — `nightly` was just a redundant single-arch tag. But now, both the amd64 and arm64 matrix jobs push `clp-package:nightly` without an arch suffix (the suffix only lands on the last tag in the newline-delimited output), so the second one to finish overwrites the first. The `nightly` tag becomes a single-arch image determined by a race condition.
> 
> I'd suggest fixing this at the root by providing explicit tag rules that omit the `type=schedule` rule, since nightly images don't appear to be an intentional feature:
> 
> ```yaml
> - name: "Extract GitHub Metadata"
>   id: "extract-gh-meta"
>   uses: "docker/metadata-action@902fa8ec7d6ecbf8d84d538b9b233a880e428804"
>   with:
>     images: "${{inputs.image_registry}}/${{steps.sanitization.outputs.REPOSITORY}}\
>       /${{steps.compute-meta.outputs.IMAGE_NAME}}"
>     tags: |
>       type=ref,event=branch
>       type=ref,event=pr
>       type=ref,event=tag
> ```
> 
> This produces exactly one tag per event type (e.g., `main` on push, `pr-N` on PRs), so the existing `-${{inputs.arch}}` suffix works correctly without changes, and no unwanted `nightly` images get published. If nightly images are wanted in the future, the `flavor` input with `suffix=-${{inputs.arch}},onlatest=true` should be used instead.

i did end up specifying the flavour but in a different way you described. the validation section in the PR description has been updated to cover more archs and also checks the manifests

junhaoliao committed:
fix(ci): Create multi-arch manifest for nightly tag on scheduled runs

The manifest job only created a manifest for the branch name tag (e.g.
`main`), but not for the `nightly` tag produced by `docker/metadata-action`'s
`type=schedule` rule. On scheduled runs, both per-arch `nightly-amd64` and
`nightly-arm64` images were pushed, but no multi-arch `nightly` manifest
was created, leaving `nightly` as a stale single-arch image.

junhaoliao committed:
Merge branch 'main' into multi-arch-docker
